File: coffeespell.py
import requests
from bs4 import BeautifulSoup
import datetime

STORENAME = 'COFFEESPELL'

def checkInt(i):
    try:
        int(i)
        return True
    except:
        return False


def conquerPage(section, url):
    itemList = []
    soup = BeautifulSoup(requests.get(url).text, 'html.parser')
    article = soup.find_all("li", {"class": '-qHwcFXhj0'})
    for a in article:
        item = {}
        # 구분 : 원두, 생두 등
        item["section"] = section
        # 이미지
        item['img'] =a.find("img", {"class": "_25CKxIKjAk"})['src']
        # 상품명
        prodName = a.find('strong', {'class': 'QNNliuiAk3'}).text.replace(',', '')
        item['product name'] = prodName
        
        item['weight'] = 0
        for n in prodName.replace('-', ' ').replace('_', ' ').split(' '):
            if 'kg' in n or 'Kg' in n or 'KG' in n:
                item['weight'] = int(n.replace('kg', '').replace('Kg', '').replace('KG', ''))*1000
            elif '0g' in n or '0G' in n:
                item['weight'] = n.replace('0G', '0').replace('0g', '0')
        print(item['weight'])
        # 가격
        item['price'] = a.find('span', {'class': 'nIAdxeTzhx'}).text.replace(',', '')
        # 리뷰수, 평점 : a.find('em', {'class': '_1dH1kEDaAZ'})
        [review, grade] = [0, 0]
        title = [r.text for r in a.find_all('span', {'class': '_1ah-_dNSCu'})]
        body = [r.text for r in a.find_all('em', {'class': '_1dH1kEDaAZ'})]
        if len(body) == 2:
            [review, grade] = body
        elif len(body) == 1:
            if title[0] == '리뷰':
                [review, grade] = [body[0], 0]
            else:
                [review, grade] = [0, body[0]]
        item['review'] = review
        item['grade'] = grade

        # 품절/재고 여부
        item['canBuy'] = 'N' if a.find('span', {'class': '_3Btky8fCyp'}) else 'Y'
        # 상세 설명
        item['detail'] = a.find('p', {'class', 'vChbm1yu9U'}).text.replace(',', '') if a.find('p', {'class', 'vChbm1yu9U'}) else 0
        itemList.append(item)
    return itemList

def conqureSection(section, url):
    itemList = []
    soup = BeautifulSoup(requests.get(url).text, 'html.parser')
    pagination = soup.find('div', {'class': '_1HJarNZHiI _2UJrM31-Ry'})
    p = sorted(filter(checkInt, [a.text for a in pagination.find_all('a')]), reverse=True)
    urlParam = int(p[0])

    for ind in range(1, urlParam + 1):
        eachUrl = url + '?cp=' + str(ind)
        itemList += conquerPage(section, eachUrl)
    return itemList

def save(obj):
    f = open(datetime.datetime.now().strftime('%Y.%m.%d_%H%M') + '_' + obj['storename'] + ".csv", "a")
    for item in obj['itemList']:
        f.write("{}, {}, {}, {}, {}, {}, {}, {}, {}, {}, {}\n".format(obj['storename'], obj['today'], item['section'], item['img'], item['product name'], item['weight'], item['price'], item['review'], item['grade'], item['canBuy'], item['detail']))
    f.close()

def main():
    obj = {"storename" : STORENAME }
    obj["today"] = datetime.datetime.now().strftime('%Y-%m-%d %H:%M:%S')
    obj["url"] = 'https://smartstore.naver.com/coffeespell'

    itemList = []
    url = 'https://smartstore.naver.com/coffeespell/category/661f22e827554c198b452d437c66dc5e' # 생두
    itemList += conqureSection('생두', url)
    url = 'https://smartstore.naver.com/coffeespell/category/ff37e49dadcc4f17b1148766d247b2bb' # 원두
    itemList += conqureSection('원두', url)
    obj["itemList"] = itemList
    # save(obj)


main()

File: coffeespell.js
import axios from "axios";
import cheerio from "cheerio";
import save from "./handleFile.js";
import inspectTitle from "./util.js";

const STORENAME = 'COFFEESPELL'

async function getItemList(section, url) {
    let itemList = []
    let soup = await axios.get(url)
    let $ = cheerio.load(soup.data)

    for(let article of $('li.-qHwcFXhj0')) {
        let item = new Map()
        item.set('section', section)
        item.set('img', $(article).find('img._25CKxIKjAk').attr('src'))
        let prodName = $(article).find('strong.QNNliuiAk3').text().replace(',', ' ').replace('_', ' ').replace('(', ' ').replace(')', ' ').replace('[', ' ').replace(']', ' ')
        item.set('product name', prodName)
        item.set('price', parseInt($(article).find('span.nIAdxeTzhx').text().replace(',', '')))

        let [review, grade] = [0, 0]
        let body = $(article).find('em._1dH1kEDaAZ')
        if(body.length == 2) {
            [review, grade] = $(body).map((i, el) => $(el).text()).toArray()
        } else if(body.length == 1) {
            if($(article).find('span._1ah-_dNSCu').text() == '리뷰') {
                review = $(body).text()
            } else {
                grade = $(body).text()
            }
        }
        item.set('review', review)
        item.set('grade', grade)

        item.set('canBuy', $(article).find('span._3Btky8fCyp').length > 0 ? "N" : "Y")
        item.set('detail', ($(article).find('p.vChbm1yu9U').text() || '').replace(',', ' '))
        item.set('extra', inspectTitle(prodName))
        itemList.push(item)
    }
    return itemList
}

async function getPageList(url) {
    let soup = await axios.get(url)
    let $ = cheerio.load(soup.data)
    
    let re = new RegExp(/[^0-9]/g)
    let maxPage = $('div._1HJarNZHiI._2UJrM31-Ry')
                        .children('a')
                        .map((i, el) => re.test($(el).text()) ? null : parseInt($(el).text()))
                        .toArray()
                        .sort((a,b) => b-a)[0]

    return Array.from(Array(maxPage), (_,i) => url + "?cp=" + (i+1))                    
}

async function init() {
    let obj = new Map()
    obj.set("storename", STORENAME)
    let now = new Date();
    obj.set("today", now.getFullYear() + "." + (now.getMonth() < 9 ? "0" : "") + (now.getMonth() + 1) + "." + (now.getDate() < 10 ? "0" + now.getDate() : now.getDate()))
    obj.set("url", 'https://smartstore.naver.com/coffeespell')

    let 생두pages = await getPageList('https://smartstore.naver.com/coffeespell/category/661f22e827554c198b452d437c66dc5e')
    for(let page of 생두pages) {
        save(obj, await getItemList('생두', page))
    }

    let 원두pages = await getPageList('https://smartstore.naver.com/coffeespell/category/ff37e49dadcc4f17b1148766d247b2bb')
    for(let page of 원두pages) {
        save(obj, await getItemList('원두', page))
    }

}

init();
